### PR TITLE
8273380: ARM32: Default to {ldrexd,strexd} in StubRoutines::atomic_{load|store}_long

### PR DESCRIPTION
**Pull request by shipilev**
Current ARM32 is one of few remaining uses of `os::is_MP`, the rest is removed by JDK-8188764. There are some interesting bugs in OS/libc that might give incorrect `os::is_MP` sometimes, e.g. in containers. Instead of risking it, we can default to {ldrexd,strexd} for ARMv7 (which always have them), and leave the `os::is_MP` path for ARMv6 (for which this is the only remaining way to load the 64-bit long). 

@mychris, you might want to take a look and do light performance testing for it?

Additional testing:
 - [x] Linux ARM32 cross-compiled build completes

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273380](https://bugs.openjdk.java.net/browse/JDK-8273380): ARM32: Default to {ldrexd,strexd} in StubRoutines::atomic_{load|store}_long

### Reviewers
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**) ⚠️ Review applies to 24b814c8efa7454a0fa11bddf4975b47efac73f3
 * [Dmitry Samersoff](https://openjdk.java.net/census#dsamersoff) (@dsamersoff - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5379/head:pull/5379` \
`$ git checkout pull/5379`

Update a local copy of the PR: \
`$ git checkout pull/5379` \
`$ git pull https://git.openjdk.java.net/jdk pull/5379/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5379`

View PR using the GUI difftool: \
`$ git pr show -t 5379`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5379.diff">https://git.openjdk.java.net/jdk/pull/5379.diff</a>

</details>

### CODE REVIEW

**Comment by shipilev** [comment]
I guess there are no takers for ARM code :)

**Review by dean-long — APPROVED** [review]
Seems good to me.  Maybe another recent arm32 contributor could also look at this?
@mychris @bulasevich @dsamersoff @TheRealMDoerr

**Comment by bulasevich** [comment]
Good change! thank you

File: src/hotspot/cpu/arm/stubGenerator_arm.cpp [hunk 1]
```diff
@@ -635,13 +635,15 @@ class StubGenerator: public StubCodeGenerator {
...
     Register src       = R0;
 
-    if (!os::is_MP()) {
-      __ ldmia(src, RegisterSet(result_lo, result_hi));
-      __ bx(LR);
-    } else if (VM_Version::supports_ldrexd()) {
+    if (VM_Version::supports_ldrexd()) {
       __ ldrexd(result_lo, Address(src));
       __ clrex(); // FIXME: safe to remove?
       __ bx(LR);
```
Comment (dsamersoff) [inline src/hotspot/cpu/arm/stubGenerator_arm.cpp:641]: bx(LR) is common for all 3 branches, so it might be better to move it out

Comment (shipilev) [reply src/hotspot/cpu/arm/stubGenerator_arm.cpp]: It seems that `bx(LR)` near `stop` is redundant. But I don't see why not move it outside the block. See new commit.

Comment (shipilev) [reply src/hotspot/cpu/arm/stubGenerator_arm.cpp]: @dsamersoff, are you happy with new version?

Comment (dsamersoff) [reply src/hotspot/cpu/arm/stubGenerator_arm.cpp]: Yes

**Review by dsamersoff — APPROVED** [review]

**Comment by shipilev** [comment]
Thank you!

/integrate

**Pull request closed without merging**